File: Coding Problems & My Solutions/3Sum/3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        if (nums == null || nums.length < 3){
            return ans;
        }
        Arrays.sort(nums);
        for (int i = 0; i < nums.length-2; i++){
            if (i != 0 && nums[i] == nums[i-1]){
                continue;
            }
            int z = nums[i];
            int a = i+1;
            int b = nums.length-1;
            while (a < b){
                int x = nums[a];
                int y = nums[b];
                if (x + y + z == 0){
                    List<Integer> l = new ArrayList<>();
                    l.add(z);
                    l.add(x);
                    l.add(y);
                    ans.add(l);
                }
                if (x + y + z > 0){
                    int currentY = y;
                    while(currentY == nums[b] && b > a){
                        b--;
                    }
                } else {
                    int currentX = x;
                    while (currentX == nums[a] && a < b){
                        a++;
                    }
                }
            }
        }
        return ans;
    }
}

File: Coding Problems & My Solutions/3Sum/3Sum.swift
class Solution {
    func threeSum(_ nums: [Int]) -> [[Int]] {
        var ans: [[Int]] = []
        var nums = nums.sorted()
        var i = 0
        while i < nums.count - 2 {
            if i != 0 && nums[i] == nums[i-1] {
                i += 1
                continue
            }
            var j = i + 1
            var k = nums.count - 1
            while j < k {
                var sum = nums[i] + nums[j] + nums[k]
                if sum == 0 {
                    var sol: [Int] = [nums[i], nums[j], nums[k]]
                    ans.append(sol)
                    repeat {
                        j += 1
                    } while j < k && nums[j] == nums[j-1]
                }
                else if sum < 0 {
                    j += 1
                } else {
                    k -= 1
                }
            }
            i += 1
        }
        return ans
    }
}